File: EmailUi/src/components/messagesService.js
import { useFetch } from "./fetch"
import { ref } from 'vue'
export function messagesService() {
  const currentMessages = ref()
  const { data, error, post, get, loading, postFile } = useFetch()

  const getQuery = (query, from, to) => {
    const must = []

    if (!query) {
      must.push({ match_all: {} })
    } else {
      must.push({ query_string: { query } })
    }

    const payload = {
      query: { bool: { must } }, from, to, sort: ["-Date"]
    }
    return payload
  }


  const getMessages = (query, callback, from = 0, to = 2, ) => {

    const searchParams = new URLSearchParams({
      from, to, search: query,
    })
    get(
        `http://localhost:8000/messages?${searchParams}`,
        callback
    )
  }

  const sumMessages = (func) => {
    return function(){
      const newMessages = func()
      if (newMessages) {
        currentMessages.value = currentMessages.value.concat(callback(newMessages))
        return newMessages
      }
    }

  }

  const addMessages = (query, from = 0, to = 0, callback) => {
    const payload = getQuery(query, from, to)
    post('http://localhost:4080/es/Messages/_search', payload, sumMessages(callback))

  }

  const uploadFile = (file, callback) => {
    const form = new FormData()
    form.append("file", file)
    postFile('http://localhost:8000/import_file', form, callback)

  }

  return {
    getMessages,
    addMessages,
    uploadFile,
    currentMessages,
    data,
    loading,
    error
  }
}

File: messages_service.go
package main

type SyncSearchBool struct {
	Must []map[string]map[string]string `json:"must"`
}

type SyncSearchQuery struct {
	Bool SyncSearchBool `json:"bool"`
}

type ZyncSearchPayload struct {
	From  int             `json:"from"`
	To    int             `json:"to"`
	Query SyncSearchQuery `json:"query"`
	Sort  []string        `json:"sort"`
}

type MessagesService struct{}

func (s *MessagesService) createPayload(
	search string,
	order []string,
	from int,
	to int,
) ZyncSearchPayload {
	boolData := SyncSearchBool{
		Must: []map[string]map[string]string{
			{"match_all": map[string]string{}},
		},
	}
	if search != "" {
		boolData = SyncSearchBool{
			Must: []map[string]map[string]string{
				{"query_string": map[string]string{
					"query": search,
				}},
			},
		}
	}

	payload := ZyncSearchPayload{
		From: from,
		To:   to,
		Query: SyncSearchQuery{
			Bool: boolData,
		},
		Sort: order,
	}
	return payload
}

func (s *MessagesService) getMessages(payload []byte) ([]byte, error) {
	return doPost(ZYNCSARCH_URL+"/es/Messages/_search", payload)
}
